sdk/core/azure-core/azure/core/tracing/_inference_api_instrumentor_impl.py

```diff
@@ -0,0 +1,379 @@
+        if span_impl_type is None:
+            return function(*args, **kwargs)
+
+        span_name = get_function_and_class_name(function, *args)
+        span = span_impl_type(name=span_name, kind=SpanKind.INTERNAL)
+        try:
+            # tracing events not supported in azure-core-tracing-opentelemetry
+            # so need to access the span instance directly
+            with span_impl_type.change_context(span.span_instance):
+                _add_request_span_attributes(span, span_name, kwargs)
+                result = await function(*args, **kwargs)
+                if kwargs.get("stream") is True:
+                    return _wrapped_stream(result, span)
+                _add_response_span_attributes(span, result)
+
+        except Exception as exc:
+            _set_attributes(span, ("error.type", exc.__class__.__name__))
+            span.finish()
+            raise
+
+        span.finish()
+        return result
+
+    return inner
+
+
+def inject_async(f, trace_type, name):
+    wrapper_fun = _trace_async_function(f)
+    wrapper_fun._original = f
+    return wrapper_fun
+
+
+def inject_sync(f, trace_type, name):
+    wrapper_fun = _trace_sync_function(f)
+    wrapper_fun._original = f
+    return wrapper_fun
+
+
+def _inference_apis():
+    sync_apis = (
+        ("azure.ai.inference", "ChatCompletionsClient", "complete", TraceType.INFERENCE, "inference_chat_completions_complete"),
+    )
+    async_apis = ()
+    return sync_apis, async_apis
+
+
+def _inference_api_list():
+    sync_apis, async_apis = _inference_apis()
+    yield sync_apis, inject_sync
+    yield async_apis, inject_async
+
+
+def _generate_api_and_injector(apis):
+    for apis, injector in apis:
+        for module_name, class_name, method_name, trace_type, name in apis:
+            try:
+                module = importlib.import_module(module_name)
+                api = getattr(module, class_name)
+                if hasattr(api, method_name):
+                    yield api, method_name, trace_type, injector, name
+            except AttributeError as e:
+                # Log the attribute exception with the missing class information
+                logging.warning(
+                    f"AttributeError: The module '{module_name}' does not have the class '{class_name}'. {str(e)}"
+                )
+            except Exception as e:
+                # Log other exceptions as a warning, as we're not sure what they might be
+                logging.warning(f"An unexpected error occurred: {str(e)}")
+
+
+def available_inference_apis_and_injectors():
+    """
+    Generates a sequence of tuples containing Inference API classes, method names, and
+    corresponding injector functions.
+
+    Yields:
+        Tuples of (api_class, method_name, injector_function)
+    """
+    yield from _generate_api_and_injector(_inference_api_list())
+
+
+def _inject_inference_api(enable_content_tracing: bool = False):
```

Comment:
I know we're still discussing where this code should live, but rather than monkey-patch like this, I think it's a better practice to use `wrapt` (what most of the opentelemetry-python-contrib packages use). We've seen issues with monkey-patching (e.g. https://github.com/open-telemetry/opentelemetry-python-contrib/issues/2364)

Reply:
after discussions going with monkey-patching for now, creating a task for investigating wrapt usage later